### DISCUSSION TITLE
SD card example only works once

### DISCUSSION THREAD

**Discussion by Rumidom** [micropython/micropython#14108] [ESP32] [answered]
hi I'm testing the spi [sdcard driver](https://github.com/micropython/micropython-lib/blob/master/micropython/drivers/storage/sdcard/sdcard.py) with the following example on a ESP32 S3 and it works great, however it only works once. If I try to run it a second time using thonny I get "OSError: no SD card", even after machine.reset() I get the same error but if I unplug the USB cable and plug it again the example works, not really sure what is going on maybe Its a issue with SDCard object already existing?

```python
import sdcard, os
from machine import Pin,SPI,SoftSPI,I2S,I2C,UART

spi  = SPI(2, sck=Pin(13), mosi=Pin(11), miso=Pin(12))
spi.init(baudrate=4000000, polarity=0, phase=0)

sd = sdcard.SDCard(spi, Pin(4))

vfs = os.VfsFat(sd)
os.mount(vfs, "/sd")

with open("/sd/twolines.txt", "w") as f:
        n = f.write("this is the first line\n this is the second")  # one block
        print(n, "bytes writen")

with open("/sd/twolines.txt", "r") as f:
        result = f.read()
        print(result)
            
os.umount('/sd')
```

another issue is that os.listdir() is broken with this driver, I always get the following error:
```
>>> os.mount(vfs, "/sd")
>>> os.chdir("/sd")
>>> os.listdir()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/lib/sdcard.py", line 230, in readblocks
OSError: [Errno 5] EIO
```

**Comment by mendenm** [comment]
Could you please try my update to the SDcard code, in PR765, https://github.com/micropython/micropython-lib/pull/765
and see if it is better.  I did quite  bit of work on the code.

**Reply by Rumidom** [reply]
hi I did notice the original driver is 9 years old, thanks for working on it. but unfortunately I got a similar behavior, the code works only once and  os.listdir() is broken giving me a different error this time  :
```
>>> os.listdir("sd")
MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 17, in <module>
  File "sdcard2.py", line 320, in readblocks
  File "sdcard2.py", line 244, in readinto
OSError: [Errno 116] ETIMEDOUT: read timeout
```

the code I tested (ESP32 S3): 
```python
import sdcard2, os
from machine import Pin,SPI,SoftSPI,I2S,I2C,UART
import utime
#from crc16 import crc16

spi  = SPI(2, 24_000_000,sck=Pin(13), mosi=Pin(11), miso=Pin(12))

spi.init(baudrate=24_000_000, polarity=0, phase=0)

sd = sdcard2.SDCard(spi=spi, cs=Pin(4, Pin.OUT), baudrate=24_000_000, crc16_function=None)

vfs = os.VfsFat(sd)
os.mount(vfs, "/sd")

with open("/sd/twolines.txt", "w") as f:
    n = f.write("this is the first line\n this is the second")  # one block
    print(n, "bytes writen")
utime.sleep_ms(100)

with open("/sd/twolines.txt", "r") as f:
    result = f.read()
    print(result)
utime.sleep_ms(100)   
```

**Comment by mendenm** [comment]
Have you tried a lower baud rate, to make sure that there isn't something that cannot run at 24 MHz?  That  should be legal for any SD card, but maybe there is something wrong.  Have you tried s different SD card?  Can you test it by putting the code in main.py, and doing a bunch of operations without Thonny involved, just in case it is somehow breaking the SPI?  I like Thonny, but its REPL interface does sometimes get in the way.

**Reply by Rumidom** [reply]
I tried different boundrates with no success, I noticed that the driver lowers  the boundrate to 100000 on init, I also tried commenting that but didnt work. I'll try a different sd card next, but I'm leaning towards modifying the wiring and using  [machine.SDcard ](https://docs.micropython.org/en/latest/library/machine.SDCard.html) instead of SPI, for now I'll just use the flash memory and once I get the next iteration of my PCB I'll try implementing  SDCard suport

**Comment by scruss** [comment]
How is the SD card attached?

**Reply by Rumidom** [reply]
a micro SDcard slot on a custom PCB, with a 100nf decoupling capacitor, do you think it could be a wiring issue? since it consistently works just after connecting the USB cable I discarded that possibility

**Comment by Rumidom** [comment, marked as answer]
I fixed the issue by changing the wiring to (sck = 39, miso=21, mosi = 42, cs =15), i think the problem was because i had a LCD on the same SPI as the sdcard, I have them separeted now and both drivers work
